### ISSUE TITLE
`StackOverflowException` - `WhatsNewPanel.cs`

### ISSUE THREAD

**Issue by aubergine10** [CitiesSkylinesMods/TMPE#1314]
> More important will be fixing `StackOverflowException` (one of the last workshop comments - Stable), log: <https://onedrive.live.com/?authkey=%21AG5WxpzcRVvkG9o&cid=7989F98A9950A030&id=7989F98A9950A030%2139983&parId=root&o=OneUp>

_Originally posted by @krzychu124 in https://github.com/CitiesSkylinesMods/TMPE/issues/1312#issuecomment-1021316872_

**Comment by krzychu124** [comment]
It looks like it is TM:PE related but I can't reproduce it

<details> 
<summary>part of stacktrace </summary>

```c#

 at ColossalFramework.UI.UIComponent.OnSizeChanged () [0x00000] in <filename unknown>:0 
  at ColossalFramework.UI.UIComponent.set_size (Vector2 value) [0x00000] in <filename unknown>:0 
  at ColossalFramework.UI.UILabel.Invalidate () [0x00000] in <filename unknown>:0 
  at ColossalFramework.UI.UIComponent.OnSizeChanged () [0x00000] in <filename unknown>:0 
  at ColossalFramework.UI.UIComponent.set_size (Vector2 value) [0x00000] in <filename unknown>:0 
  at ColossalFramework.UI.UILabel.Invalidate () [0x00000] in <filename unknown>:0 
  at ColossalFramework.UI.UIComponent.OnSizeChanged () [0x00000] in <filename unknown>:0 
  at ColossalFramework.UI.UIComponent.set_size (Vector2 value) [0x00000] in <filename unknown>:0 
UnityEngine.GameObject:Internal_AddComponentWithType(Type)
UnityEngine.GameObject:AddComponent(Type)
UnityEngine.GameObject:.ctor(String, Type[])
ColossalFramework.UI.UIView:AddUIComponent(Type)
TrafficManager.UI.WhatsNew.WhatsNew:OpenModal()
<AddButton>c__AnonStorey4:<>m__0(UIComponent, UIMouseEventParameter)
ColossalFramework.UI.UIComponent:OnClick(UIMouseEventParameter)
ColossalFramework.UI.UIButton:OnClick(UIMouseEventParameter)
ColossalFramework.UI.MouseHandler:ProcessInput(IInputTranslator, Ray, UIComponent, Boolean)
ColossalFramework.UI.UIInput:ProcessMouseInput()
ColossalFramework.UI.UIInput:Update()

```
</details>

**Comment by krzychu124** [comment]
As a blind fix we should limit amount of text in the changelog to reduce chance of Stackoverflow. I have no idea what to do to break it...

**Comment by aubergine10** [comment]
Is it only one user getting that problem?

**Comment by krzychu124** [comment]
4 or 5? - Stable workshop comments section

**Comment by aubergine10** [comment]
And it's only whey they click the `What's New` button?

**Comment by krzychu124** [comment]
Yes, or.. when they didn't and click on TMPE main menu for the first time 😕 I'll generate 10x more text and try again. If there is an issue with component count it will appear at some point. Alternatively we can disable auto open for a while, maybe it's just interference with broken mod (log above contains exceptions so I'm not sure it's not related since Klytes TLM creates tons of new ui components in game)?

**Comment by aubergine10** [comment]
Do they have FPS Booster? Is it possible that the new font stuff causes some intermittent sizing issues while UI components are being added?

**Comment by krzychu124** [comment]
It looks like they don't. I also tested it without any other mod and I just can't break it. I've tested it with total of 32 changelogs, complete vanilla and opening panel takes ~10 seconds. Still no crash/exception...

**Comment by krzychu124** [comment]
Ok... I think I've managed to trigger CTD 😄 No errors, just poof and closed. Let's try again this time with debugger 😂

**Comment by aubergine10** [comment]
Might need to stop auto layout while adding the UI components, then start autolayout once everything is added?

**Comment by krzychu124** [comment]
@aubergine10 I think we forgot to test case if there is no LastSeen entry in the config -> does not show automatically and when I open it manually it shows everything. Maybe we should add default value to config (11.6.4.0 or any)?

**Comment by aubergine10** [comment]
In that situation we should probably limit to say, max 5 changelogs applicable to the branch the user is using? I was thinking of doing something along those lines. 

IMO release the 11.6.4.1 as that will keep the 'train speeds missing' people happy and will give us some indication of wheather the recent updates to WhatsNew are solving the stack overflow.

We can add the max 5 changelogs thing (or whatever is decided) later in 11.6.5.0 ?

**Comment by krzychu124** [comment]
> We can add the max 5 changelogs thing (or whatever is decided) later in 11.6.5.0 ?

Yup, good idea... Releasing

**Comment by krzychu124** [comment]
More reports, hardcoding what's new version in globalconfig does not fix issue with crash on click at TM:PE main icon after loading savegame.

[TMPE (2).log](https://github.com/CitiesSkylinesMods/TMPE/files/7964769/TMPE.2.log)
[output_log.txt](https://github.com/CitiesSkylinesMods/TMPE/files/7964768/output_log.txt)

I need to redesign code to eliminate autolayout or autosize of labels since is seems to be the main source of stack overflow

I think we should temporarily disable auto open (and auto game pause) of What's new panel until I fix issues

**Comment by aubergine10** [comment]
Pushed a PR to disable the what's new on tmpe button click https://github.com/CitiesSkylinesMods/TMPE/pull/1336

**Comment by krzychu124** [comment]
I asked MacSergey, he also faced that issue before and told me what situation can trigger the issue. 
I'll try to trigger it then redesign the code to fix the problem but for now, thanks for disabling it!

**Comment by aubergine10** [comment]
Yup, I was looking through some of his UI code the other day and all of it has stuff that controls the autolayout stuff.

**Comment by DaEgi01** [comment]
> More reports, hardcoding what's new version in globalconfig does not fix issue with crash on click at TM:PE main icon after loading savegame.
> 
> [TMPE (2).log](https://github.com/CitiesSkylinesMods/TMPE/files/7964769/TMPE.2.log) [output_log.txt](https://github.com/CitiesSkylinesMods/TMPE/files/7964768/output_log.txt)
> 
> I need to redesign code to eliminate autolayout or autosize of labels since is seems to be the main source of stack overflow
> 
> I think we should temporarily disable auto open (and auto game pause) of What's new panel until I fix issues

I remember running into perf issues with autolayout / autosize but I didnt bother looking into it. instead i just went for manual layout since most of the time of most mods UI is static anyway xD

**Comment by aubergine10** [comment]
![image](https://user-images.githubusercontent.com/1386719/154830719-49e57c7b-0be3-4bc0-86df-949d2149f3f1.png)

**Comment by kianzarrin** [comment]
can we ask them to see if they can reproduce this problem using minimal mod set?

**Comment by aubergine10** [comment]
I think we just need to publish 11.6.5.0 to STABLE.

**Comment by krzychu124** [comment]
I would ask first if TEST fix the issue (just test in main menu, no need to load savegame), if not then we will have to push another hotfix turning off that feature again

**Comment by krzychu124** [comment]
No more reports about game crashes when opening What's New panel.
